server/src/pages/shot/editor-history.js

```diff
@@ -0,0 +1,118 @@
+const { Selection } = require("../../../shared/selection");
+
+exports.EditorHistory = class {
+  constructor(devicePixelRatio) {
+    this.beforeEdits = [];
+    this.afterEdits = [];
+    this.devicePixelRatio = devicePixelRatio;
+  }
+
+  push(canvas, area, recordType) {
+    const record = new EditRecord(
+      canvas,
+      area,
+      this.devicePixelRatio,
+      recordType
+    );
+    this.beforeEdits.push(record);
+    this.afterEdits = [];
+  }
+
+  pushDiff(canvas, area) {
+    this.push(canvas, area, RecordType.DIFF);
+  }
+
+  pushFrame(canvas, area) {
+    this.push(canvas, area, RecordType.FRAME);
+  }
+
+  canUndo() {
+    return !!this.beforeEdits.length;
+  }
+
+  undo(canvasBeforeUndo) {
+    if (!this.canUndo()) {
+      return null;
+    }
+
+    return this._replay(canvasBeforeUndo, this.beforeEdits, this.afterEdits);
+  }
+
+  canRedo() {
+    return !!this.afterEdits.length;
+  }
+
+  redo(canvasBeforeRedo) {
+    if (!this.canRedo()) {
+      return null;
+    }
+
+    return this._replay(canvasBeforeRedo, this.afterEdits, this.beforeEdits);
+  }
+
+  _replay(canvasBeforeChange, from, to) {
```

Comment:
I'm a bit confused by this part of the code. If I understand correctly, in the `push` method above, each time the image is edited, an `EditRecord` is appended to `this.beforeEdits`. Why, then, does this method need to create a new `EditRecord` each time undo or redo is pressed? I wonder if it would be more performant to have just one array of edits, and, when undo or redo is pressed, increment or decrement an index cursor into that existing array of `EditRecords`.